### PR TITLE
Adding software implementing TEA

### PR DESCRIPTION
**Pull request by kornelu123**
Instead of implementing RSA encryption algorithm, which is unpractical due to necessity of finding big prime numbers which is taking very long time I went with Tiny Encryption Algorithm, which do not have those flaws.

### CODE REVIEW

**Comment by kornelu123** [comment]
Mmmmm, what a beautiful piece of Object oriented programming ;]. :face_exhaling:

File: tea.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      const uint32_t u32_datasize = 
+        u8_datasize % TEA_DATABLOCK_SIZE == 0 ? 
+        u8_datasize/TEA_DATABLOCK_SIZE : 
+        (u8_datasize/TEA_DATABLOCK_SIZE + 1)*2;
+
+      this->data = (uint32_t *)calloc(u32_datasize, sizeof(uint32_t));
+      memcpy(this->data, data, u8_datasize);
+
+      this->data_size = u32_datasize;
+    }
```
Comment (husxk) [inline tea.cpp]: Can't we just do
`tea_ctx_t(uint32_t* data, uint32_t u32_datasize)`?\
Then we do not need to calculate `u32_datasize`.

Also, then we can just do `std::memcpy(&(this->data), (void*)data, u32_datasize * 4);`

File: tea.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    ~tea_ctx_t()
+    {
+      free(this->data);
+    }
+
+    void
+    print_data(void)
+    {
+      for(int i=0; i<this->data_size; i++){
+        printf("%08X", this->data[i]);
```
Comment (husxk) [inline tea.cpp]: `printf("%08X ", this->data[i]);`?
The whitespace will make output more clear, no?
It's up to you whether change it or not.

File: tea.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        sum += TEA_DELTA;
+        val_0 += ((val_1<<4) + key[0]) ^ (val_1 + sum) ^ ((val_1>>5) + key[3]);
+        val_1 += ((val_0<<4) + key[2]) ^ (val_0 + sum) ^ ((val_0>>5) + key[1]);
+      }
+    
+      data[index] = val_0;
+      data[index+1] = val_1;
+    }
+
+    inline void
```
Comment (husxk) [inline tea.cpp]: FYI: all functions here (class definition) are `inline` implicitly
([Inline explained at cppreference](https://en.cppreference.com/w/cpp/language/inline))

**Review by husxk — CHANGES_REQUESTED** [review]
It looks good, however can we break it to .h and .cpp files?
We can make `source/` and `include/` directories.
(You can add .h and .cpp file to your changes in one commit,
and restructure the project files in another one)
Do we want to leave `main.cpp`  in the main directory? What do you think?

Also, shouldn't we update CMakeLists.txt accordingly?

**Comment by kornelu123** [comment]
> It looks good, however can we break it to .h and .cpp files? 

Yeah we can, I'm not a fan of it but if You want then let's make it be.

> Also, shouldn't we update CMakeLists.txt accordingly?

Due to not having header file, just including .cpp file works.

> We can make `source/` and `include/` directories. (You can add .h and .cpp file to your changes in one commit, and restructure the project files in another one) Do we want to leave `main.cpp` in the main directory? What do you think?

TBH i think this is a small project, so we do not need the src/ and inc/ dirs. But it's your repo so I leave it up to you.

File: tea.cpp (hunk 3 above)
Comment (kornelu123) [reply tea.cpp]: GTK, i think i would like to still explicitly write inline, what are your thoughts?

Comment (husxk) [reply tea.cpp]: I don't mind having it explicitly.

**Comment by husxk** [comment]
> TBH i think this is a small project, so we do not need the src/ and inc/ dirs. But it's your repo so I leave it up to you.

I agree that it is a small project, so we can leave it as it is. If we get a ton of files, we will separate .cpp and .h

File: main.cpp [hunk 4]
```diff
@@ -1,4 +1,6 @@
 #include "mt_rng.hpp"
+#include "tea.cpp"
+#include "pattern_map.cpp"
```
Comment (husxk) [inline main.cpp]: there is no such a file

File: main.cpp [hunk 5]
```diff
@@ -1,4 +1,6 @@
 #include "mt_rng.hpp"
+#include "tea.cpp"
```
Comment (husxk) [inline main.cpp]: shouldn't we `#include "tea.hpp"`?

**Review by husxk — CHANGES_REQUESTED** [review]
Shouldn't we change CMakeLists.txt appropriately?

**Comment by husxk** [comment]
Looks good, have you tested it?

**Comment by kornelu123** [comment]
yup

**Pull request merged**